tools: fix integer scaling for metis/scotch

hhirtz opened the pull request:
The idea is that we have weights
- in some [min; max] range,
- whose sum is S

and we want to scale them so
1. they fit into some integer type,
2. the output sum is below INTMAX
3. the output has no zeroes

basically we apply an affine function "y=scale*x+offset" where scale and offset depend on the max output value M:

         (M-1) * x + max - M*min
    y = -------------------------
               max - min

y=1 for x=min and y=M for x=max

M is set so that the sum of all "y" is INTMAX (N is the number of weights):

         (max - min) * INTMAX + sum(x) - max * N
    M = -----------------------------------------
                   sum(x) - min * N

cedricchevalier19 commented:
> ```math       
> M = \frac{(max - min) \times INTMAX + sum(x) - max \times N}{sum(x) - min \times N}
> ```

Be careful : $(max - min) \times INTMAX$ or $max \times N$ are likely to overflow if these computations are not done in floating point arithmetic.

hhirtz commented:
scale and offset are computed using f64s exclusively, then weights are scaled with a fused multiply-add and then converted to the output integer type.

Opened a new issue for the clippy error: #299

hhirtz commented on tools/src/lib.rs:
```diff
@@ -839,15 +848,46 @@ where
-        .unwrap()
-        / sum.to_f64().unwrap();
-    let _1 = O::one();
+    let (min_input, max_input) = inputs.clone().flatten().minmax().into_option().unwrap();
+    if min_input == max_input {
+        return vec![O::one(); count * criterion_count];
+    }
```
Actually it might be better to have `zoom_in` return `None` when `min = max`. Then we can leave the weight/adjwgt parameter of scotch/metis as the default.

I'll upload a new patch soon.

hhirtz replied:
done

hhirtz commented:
There was indeed a possible overflow, in the computation of the criterion sums. I changed it to a compensated floating-point sum.

I also changed the way weights are scaled, `min_input` is set to 0 now, so that the ratios between weights are kept more or less the same as they were originally. For example, if all weights are close to some big number (eg `[INTMAX-3, INTMAX-2, INTMAX-1]`), it makes more sense to make them all equal than to "zoom in" to something like `[1, INTMAX/3, 2/3*INTMAX]`). Otherwise it won't represent the original weights faithfully.

Maybe there is a known way to move the inputs into integer range while keeping ratios.
